### ISSUE TITLE
Wkhtmltopdf not installed

### ISSUE THREAD

**Issue by sbglive** [bitnami/bitnami-docker-odoo#43]
Dear Support,

When clicking "print" on any documents in Odoo on a standard setup, I get the message:
"Unable to find Wkhtmltopdf on this system. The report will be shown in html.
wkhtmltopdf.org"

I believe wkhtmltopdf is not installed on the image, which is an Odoo requirement to print documents as PDF.

Is it possible to include it by default as it affects a core feature of the application ?

Thank you,
Benoit.

**Comment by juan131** [comment]
Hello Benoit,

You can follow this guide to install it:
https://docs.bitnami.com/installer/apps/odoo/#how-to-install-wkhtmltopdf

But please use `nami restart apache` instead of `sudo installdir/ctlscript.sh restart apache` to restart Apache since you're using a container.

However, I think it could be interesting to include it by default on the Docker Image. What do you think @tompizmor @jotadrilo ?

**Comment by sbglive** [comment]
Hello Juan,

Thank you for your quick reply. 
I just tried, but it seems some dependencies are not met (libXrender.so.1). I downloaded the sources, moved them to /usr/bin as suggested in the link you sent but the message is still showed in Odoo.

I tried to execute the binary from command line to see the results, I get this message:
```bash
root@b0ff7792a774:/# wkhtmltopdf
wkhtmltopdf: error while loading shared libraries: libXrender.so.1: cannot open shared object file: No such file or directory
root@b0ff7792a774:/# wkhtmltoimage
wkhtmltoimage: error while loading shared libraries: libXrender.so.1: cannot open shared object file: No such file or directory
```

Unfortunately I am not very familiar with debian.
I noted that command wget is not available on the docker image, and also had an error when extracting the sources (tar -xvf). So I downloaded the sources from the host, moved it to a shared volume with the container, then extracted it from the host, then moved it in /usr/bin from the container.

Not sure if I am doing this right - I guess your debian image is lightweight and may not include all the standard tools of a traditional debian.

Did you manage to make it work on your side ?


Thanks,
Benoit.

**Comment by juan131** [comment]
> Not sure if I am doing this right - I guess your debian image is lightweight and may not include all the standard tools of a traditional debian.

Yes.. That's basically the reason why you're facing the issue with the missing libraries... Bitnami images are based on an image called 'minideb' which is an extremely optimised image that packages only the necessary to run the application.

I honestly think we should include this tool by default on the image... Waiting for @tompizmor and @jotadrilo thoughts...

**Comment by Tofdu31** [comment]
Hello Benoit,

I have the same problem as you.
Can you explain to me how did you connect to /usr/ bin?

Thank

**Comment by carrodher** [comment]
`/usr/bin` is a standard directory on Unix-like operating systems that contains most of the executable files. You can use it as a regular directory in a Unix system, in this case, @sbglive download the sources in him local machine (in a directory mounted as a volume in the container) and move the executable to the `/usr/bin` folder.

**Comment by jotadrilo** [comment]
Hi @juan131, @sbglive 

I checked that including the required libs and binaries to support this tool would increase the image size by 80MB (~30MB compressed). However, as it is a third-party tool it would be hard to keep it updated within the image.

On the other hand, I think we can assume the overload of including the required libs so far. It would increase the size by 2MB only, which it is insignificant once compressed. We could even include `xz-utils` to allow users to untar `.tar.xz` files, so it would be possible to inherit from the `bitnami/odoo` image and install the tool as follow:

```
RUN curl -LO https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.4/wkhtmltox-0.12.4_linux-generic-amd64.tar.xz && \
    tar xf wkhtmltox-0.12.4_linux-generic-amd64.tar.xz && \
    cp wkhtmltox/bin/* /usr/local/bin/ && \
    rm -rf wkhtmltox*
```

If this is a common request from the community, we will investigate the best way of adding this tool to the image too. However, I think this would be enough for now. Note that docker-compose also allows to mount local binaries into the image. Example:

```
...
  odoo:
    image: 'bitnami/odoo:11'
    ports:
      - '80:8069'
      - '443:8071'
    volumes:
      - 'odoo_data:/bitnami'
      - /path/to/local/wkhtmltoimage:/usr/local/bin/wkhtmltoimage
      - /path/to/local/wkhtmltopdf:/usr/local/bin/wkhtmltopdf
    depends_on:
      - postgresql
```

What do you guys think?

> NOTE: The snippets above are thought to be run within an image that already includes `libxrender1` package. If you try them out, they will fail.

**Comment by juan131** [comment]
I agree with you @jotadrilo It's too much size for this Docker image. 

I'd be enough for me if we include `xz-utils` so the user can easily install `wkhtmltopdf` using a "curl-tar" combo command.

**Comment by jotadrilo** [comment]
+1
could you create an internal task to include them?

**Comment by Dynogic** [comment]
Hey guys,

We have to tell Odoo that _/usr/bin/_ is a desired _PATH_ to look in.

On the Bitnami installation, the _PATH_ variable Python references does not contain the regular system directories like: `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`

Debug printing `os.environ.get('PATH', os.defpath)` in Odoo yields `/opt/bitnami/node/bin:/opt/bitnami/odoo/node_modules/less/bin:$PATH`

Seems like the following code in Odoo doesn't resolve that _$PATH_ variable:
```python
def find_in_path(name):
    path = os.environ.get('PATH', os.defpath).split(os.pathsep)
    if config.get('bin_path') and config['bin_path'] != 'None':
        path.append(config['bin_path'])
    return which(name, path=os.pathsep.join(path))
```

Therefore, after following https://docs.bitnami.com/virtual-machine/apps/odoo/#how-to-install-wkhtmltopdf with the binaries at https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.4/wkhtmltox-0.12.4.tar.bz2 in the Docker container's _bash_...

... modify _odoo-server.conf_ by adding-in the following line:
`bin_path = /usr/bin`

Then, restart the Odoo docker container. That made it work for me!

**Is this a possible bug in the Bitnami?** Echoing _PATH_ in the Odoo container's _bash_ yields
`/opt/bitnami/python/bin:/opt/bitnami/postgresql/bin:/opt/bitnami/node/bin:/opt/bitnami/nami/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`. Some input will be great!

**Comment by juan131** [comment]
Hi @Dynogic 

I was unable to reproduce the error. It **does** retrieve the PATH when running the command you suggested. See the output below:

```
$ docker-compose exec odoo echo $PATH
$ docker-compose exec odoo python
Python 3.6.5 (default, Jun 21 2018, 10:10:01)
[GCC 6.3.0 20170516] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ.get('PATH', os.defpath)
'/opt/bitnami/python/bin:/opt/bitnami/postgresql/bin:/opt/bitnami/node/bin:/opt/bitnami/nami/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
>>> exit()
```

**Comment by Dynogic** [comment]
Hi, thanks Juan.

Using the above terminal output, I am also able to replicate it. 

However, can you do me a favour and modify `ir_actions_report.py` in the Odoo container

`vi /opt/bitnami/odoo/odoo/addons/base/ir/ir_actions_report.py`

Changing method body of `_get_wkhtmltopdf_bin` to 

```python
def _get_wkhtmltopdf_bin():
    path = os.environ.get('PATH', os.defpath)
    _logger.info('Path: %s' % path)
    return find_in_path('wkhtmltopdf')
```
Delete the Odoo log located at `/opt/bitnami/odoo/logs/odoo-server.log` and restart the container.

Then, taking a look at the contents of the log? My log indicates the _PATH_ output to be:
`2018-07-18 17:18:16,819 29 INFO bitnami_odoo odoo.addons.base.ir.ir_actions_report: Path: /opt/bitnami/node/bin:/opt/bitnami/odoo/node_modules/less/bin:$PATH`

If you're seeing a similar log (i.e. the _$PATH_ __AND__ mention of Node Less bin location that's not in aforementioned Python REPL output), then any idea to what it could be?

Cheers,
Jake.

**Comment by juan131** [comment]
Hi @carrodher 

I introduced the lines you suggested:

```
    path = os.environ.get('PATH', os.defpath)
    _logger.info('Path: %s' % path)
```

Then, I restart Odoo as shown below:

```
$ docker-compose exec odoo nami restart odoo
com.bitnami.odoo restarted
```

Finally I checked the logs:

```
$ docker-compose exec odoo cat /opt/bitnami/odoo/logs/odoo-server.log
...
2018-07-19 08:04:20,486 392 INFO bitnami_odoo odoo.addons.base.ir.ir_actions_report: Path: /opt/bitnami/node/bin:/opt/bitnami/odoo/node_modules/less/bin:/opt/bitnami/python/bin:/opt/bitnami/postgresql/bin:/opt/bitnami/node/bin:/opt/bitnami/nami/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
...
```

Again.. It seems to be retrieving the right path... Did you change anything on your Odoo container configuration?

**Comment by Dynogic** [comment]
Thanks @juan131,

Spinning up a new VM with Ubuntu 18.04 LTS, Docker version 18.06.0-ce, and the docker-compose YML in this repository, I was able to reproduce your log entry:
`2018-07-19 15:54:25,972 30 INFO bitnami_odoo odoo.addons.base.ir.ir_actions_report: Path: /opt/bitnami/node/bin:/opt/bitnami/odoo/node_modules/less/bin:/opt/bitnami/python/bin:/opt/bitnami/postgresql/bin:/opt/bitnami/node/bin:/opt/bitnami/nami/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`

I did not change anything in the Odoo container configuration. Something, somewhere went wrong. :unamused: Can't be bothered to investigate further right now, however.

Hopefully someone finds my aforementioned fix useful!

Thanks for the support Juan! :satisfied:

Cheers,
Jake.

**Comment by javsalgar** [comment]
Strange... However, good to know that it is fixed! :D

**Comment by petanikode-kode** [comment]
thank for sharing please anybody already figure out how to configure wkhtmltopdf on bitnami window already download it but i do not which folder be suit with that, 2 folder include and bin zip file.

please detail for put that file.

Regards,

Li

**Comment by juan131** [comment]
Hi @petanikode-kode 

> thank for sharing please anybody already figure out how to configure wkhtmltopdf on bitnami window already download it but i do not which folder be suit with that, 2 folder include and bin zip file.

Are you using a Odoo native installer for Windows or are you using the Odoo container using Docker on Windows (or any other similar container technology)?
